pkg/rbd/rbd_util.go

```diff
@@ -0,0 +1,373 @@
+	if client == nil {
+		return "", fmt.Errorf("Cannot get kube client")
+	}
+	secrets, err := client.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	secret := ""
+	for k, v := range secrets.Data {
+		if k == secretName {
+			return string(v), nil
+		}
+		secret = string(v)
+	}
+
+	return secret, nil
+}
+
+func attachRBDImage(image string, volOptions *rbdVolumeOptions) (string, error) {
+	var err error
+	var output []byte
+
+	devicePath, found := waitForPath(volOptions.pool, image, 1)
+	if !found {
+		attachdetachMutex.LockKey(string(volOptions.pool + image))
+		defer attachdetachMutex.UnlockKey(string(volOptions.pool + image))
+
+		_, err = execCommand("modprobe", []string{"rbd"})
+		if err != nil {
+			glog.Warningf("rbd: failed to load rbd kernel module:%v", err)
+		}
+
+		backoff := wait.Backoff{
+			Duration: rbdImageWatcherInitDelay,
+			Factor:   rbdImageWatcherFactor,
+			Steps:    rbdImageWatcherSteps,
+		}
+		err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+			used, rbdOutput, err := rbdStatus(image, volOptions)
+			if err != nil {
+				return false, fmt.Errorf("fail to check rbd image status with: (%v), rbd output: (%s)", err, rbdOutput)
+			}
+			return !used, nil
+		})
+		// return error if rbd image has not become available for the specified timeout
+		if err == wait.ErrWaitTimeout {
+			return "", fmt.Errorf("rbd image %s/%s is still being used", volOptions.pool, image)
+		}
+		// return error if any other errors were encountered during wating for the image to becme avialble
+		if err != nil {
+			return "", err
+		}
+
+		glog.V(1).Infof("rbd: map mon %s", volOptions.monitors)
+		// If we don't have admin id/secret (e.g. attaching), fallback to user id/secret.
+		id := volOptions.adminID
+		secret := volOptions.adminSecret
+		if id == "" {
+			id = volOptions.userID
+			secret = volOptions.userSecret
+		}
+
+		output, err = execCommand("rbd", []string{
+			"map", image, "--pool", volOptions.pool, "--id", id, "-m", volOptions.monitors, "--key=" + secret})
+		if err != nil {
+			glog.V(1).Infof("rbd: map error %v, rbd output: %s", err, string(output))
+			return "", fmt.Errorf("rbd: map failed %v, rbd output: %s", err, string(output))
+		}
+		devicePath, found = waitForPath(volOptions.pool, image, 10)
+		if !found {
+			return "", fmt.Errorf("Could not map image %s/%s, Timeout after 10s", volOptions.pool, image)
+		}
+	}
+
+	return devicePath, nil
+}
+
+func detachRBDImage(image string, volOptions *rbdVolumeOptions) error {
+	var err error
+	var output []byte
+
+	glog.V(1).Infof("rbd: unmap mon %s", volOptions.monitors)
+	// If we don't have admin id/secret (e.g. attaching), fallback to user id/secret.
+	id := volOptions.adminID
+	secret := volOptions.adminSecret
+	if id == "" {
+		id = volOptions.userID
+		secret = volOptions.userSecret
+	}
+
+	output, err = execCommand("rbd", []string{
```

Comment:
unmap doesn't need key and secret, you just need device path and run e.g `rbd unmap /dev/rbd0`

Reply:
I was hoping to use different way of unmap with using image and pool parameters to eliminate search for device. On the system with large number of volumes it could be expensive I think.